frame/node-permission/src/lib.rs

```diff
@@ -0,0 +1,352 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2019-2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Node permission moudule
+//!
+//! This module is used by the `client/service` to retrieve the current
+//! permissioned nodes.
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use sp_std::prelude::*;
+use sp_std::fmt::Debug;
+use sp_core::storage::well_known_keys;
+use sp_runtime::traits::{Member, MaybeSerializeDeserialize, MaybeDisplay};
+use frame_support::{
+    decl_module, decl_storage, decl_event, decl_error,
+    Parameter, storage, ensure,
+    weights::{DispatchClass, Weight},
+    traits::{Get, EnsureOrigin},
+};
+use codec::Encode;
+
+pub trait WeightInfo {
+    fn add_node() -> Weight;
+    fn remove_node() -> Weight;
+    fn swap_node() -> Weight;
+    fn reset_nodes() -> Weight;
+}
+
+impl WeightInfo for () {
+    fn add_node() -> Weight { 50_000_000 }
+    fn remove_node() -> Weight { 50_000_000 }
+    fn swap_node() -> Weight { 50_000_000 }
+    fn reset_nodes() -> Weight { 50_000_000 }
+}
+
+pub trait Trait: frame_system::Trait {
+    /// The event type of this module.
+	type Event: From<Event> + Into<<Self as frame_system::Trait>::Event>;
+
+    /// The maximum number of permissioned nodes that are allowed to set
+    type MaxPermissionedNodes: Get<u32>;
+
+    /// The node identifier type for the runtime.
+    type NodeId: Parameter + Member + MaybeSerializeDeserialize + Debug + MaybeDisplay
+        + Ord + Default;
+
+    /// The origin which can add a permissioned node.
+    type AddOrigin: EnsureOrigin<Self::Origin>;
+
+    /// The origin which can remove a permissioned node.
+    type RemoveOrigin: EnsureOrigin<Self::Origin>;
```

Comment:
Do we really need a special origin per operation?

Reply:
I am inclined to believe that for a pallet such as this it is better err on the side of granularity, which is to say yes I would think it's better to have a special origin per operation. Would be interesting to see what Kirill has to say about the design of this pallet.